### PR TITLE
feat(nvidia-library): update tensorrt and cudnn

### PR DESCRIPTION
**Pull request by wep21**
## Description

update nvidia library version
https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/libcudnn8_8.2.4.15-1+cuda11.4_amd64.deb
https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/libnvinfer8_8.2.4-1+cuda11.4_amd64.deb

All modules which depend on these libraries work correctly. See https://github.com/autowarefoundation/autoware.universe/issues/589.

Nvidia has recently launched a major update to TensorRT library(from 8.2.2 to 8.2.4).
https://docs.nvidia.com/deeplearning/tensorrt/release-notes/tensorrt-8.html#tensorrt-8
This update includes some bug fixes and performance improvements, so it is preferable to update the version if there is no change which breaks the existing features in autoware.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kenji-miyake** [comment]
@wep21 Could you write some reasons for updating the versions?

**Review by yukkysaito — DISMISSED** [review]

**Comment by kenji-miyake** [comment]
I'll confirm if I can update the versions by just running `./setup-dev-env.sh universe`.

**Comment by kenji-miyake** [comment]
Failed. :pleading_face: (But it seems this PR isn't related?)

```
TASK [autoware.dev_env.cuda : Install cuda-11-4] ******************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"cache_update_time": 1649300457, "cache_updated": true, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"       install 'cuda-11-4'' failed: E: Unable to correct problems, you have held broken packages.\n", "rc": 100, "stderr": "E: Unable to correct problems, you have held broken packages.\n", "stderr_lines": ["E: Unable to correct problems, you have held broken packages."], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nSome packages could not be installed. This may mean that you have\nrequested an impossible situation or if you are using the unstable\ndistribution that some required packages have not yet been created\nor been moved out of Incoming.\nThe following information may help to resolve the situation:\n\nThe following packages have unmet dependencies:\n cuda-11-4 : Depends: cuda-runtime-11-4 (>= 11.4.4) but it is not going to be installed\n             Depends: cuda-demo-suite-11-4 (>= 11.4.100) but it is not going to be installed\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "Some packages could not be installed. This may mean that you have", "requested an impossible situation or if you are using the unstable", "distribution that some required packages have not yet been created", "or been moved out of Incoming.", "The following information may help to resolve the situation:", "", "The following packages have unmet dependencies:", " cuda-11-4 : Depends: cuda-runtime-11-4 (>= 11.4.4) but it is not going to be installed", "             Depends: cuda-demo-suite-11-4 (>= 11.4.100) but it is not going to be installed"]}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=44   changed=7    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0

Failed.
```

**Comment by wep21** [comment]
> Could you write some reasons for updating the versions?

@kenji-miyake I added the rational reason to update the version in the description.

**Comment by kenji-miyake** [comment]
@wep21 Thank you! But could you write a bit more concrete differences between `8.2.2` and `8.2.4`? Since we already use the major version `8`, I guess `from 7 to 8` is irrelevant.

**Comment by kenji-miyake** [comment]
I'm sorry, I had `cuda-11-1`. lol
So I'll update it to `cuda-11-4` and `tensorrt/cudnn 8.2.2` first, and try updating them to `8.2.4`.

```
❯ dpkg -l | grep cuda
ii  cuda-11-1                                           11.1.1-1                                                       amd64        CUDA 11.1 meta-package
```

**Comment by kenji-miyake** [comment]
```
❯ dpkg -l | grep cuda
ii  cuda-11-4                                           11.4.4-1                                                       amd64        CUDA 11.4 meta-package
ii  cuda-cccl-11-4                                      11.4.122-1                                                     amd64        CUDA CCCL
ii  cuda-command-line-tools-11-4                        11.4.4-1                                                       amd64        CUDA command-line tools
ii  cuda-compiler-11-4                                  11.4.4-1                                                       amd64        CUDA compiler
ii  cuda-cudart-11-4                                    11.4.148-1                                                     amd64        CUDA Runtime native Libraries
ii  cuda-cudart-dev-11-4                                11.4.148-1                                                     amd64        CUDA Runtime native dev links, headers
ii  cuda-cuobjdump-11-4                                 11.4.120-1                                                     amd64        CUDA cuobjdump
ii  cuda-cupti-11-4                                     11.4.120-1                                                     amd64        CUDA profiling tools runtime libs.
ii  cuda-cupti-dev-11-4                                 11.4.120-1                                                     amd64        CUDA profiling tools interface.
ii  cuda-cuxxfilt-11-4                                  11.4.120-1                                                     amd64        CUDA cuxxfilt
ii  cuda-demo-suite-11-4                                11.4.100-1                                                     amd64        Demo suite for CUDA
ii  cuda-documentation-11-4                             11.4.126-1                                                     amd64        CUDA documentation
ii  cuda-driver-dev-11-4                                11.4.148-1                                                     amd64        CUDA Driver native dev stub library
ii  cuda-drivers                                        510.47.03-1                                                    amd64        CUDA Driver meta-package, branch-agnostic
ii  cuda-drivers-510                                    510.47.03-1                                                    amd64        CUDA Driver meta-package, branch-specific
ii  cuda-gdb-11-4                                       11.4.120-1                                                     amd64        CUDA-GDB
ii  cuda-libraries-11-4                                 11.4.4-1                                                       amd64        CUDA Libraries 11.4 meta-package
ii  cuda-libraries-dev-11-4                             11.4.4-1                                                       amd64        CUDA Libraries 11.4 development meta-package
ii  cuda-memcheck-11-4                                  11.4.120-1                                                     amd64        CUDA-MEMCHECK
ii  cuda-nsight-11-4                                    11.4.120-1                                                     amd64        CUDA nsight
ii  cuda-nsight-compute-11-4                            11.4.4-1                                                       amd64        NVIDIA Nsight Compute
ii  cuda-nsight-systems-11-4                            11.4.4-1                                                       amd64        NVIDIA Nsight Systems
ii  cuda-nvcc-11-4                                      11.4.152-1                                                     amd64        CUDA nvcc
ii  cuda-nvdisasm-11-4                                  11.4.152-1                                                     amd64        CUDA disassembler
ii  cuda-nvml-dev-11-4                                  11.4.120-1                                                     amd64        NVML native dev links, headers
ii  cuda-nvprof-11-4                                    11.4.120-1                                                     amd64        CUDA Profiler tools
ii  cuda-nvprune-11-4                                   11.4.120-1                                                     amd64        CUDA nvprune
ii  cuda-nvrtc-11-4                                     11.4.152-1                                                     amd64        NVRTC native runtime libraries
ii  cuda-nvrtc-dev-11-4                                 11.4.152-1                                                     amd64        NVRTC native dev links, headers
ii  cuda-nvtx-11-4                                      11.4.120-1                                                     amd64        NVIDIA Tools Extension
ii  cuda-nvvp-11-4                                      11.4.193-1                                                     amd64        CUDA Profiler tools
ii  cuda-runtime-11-4                                   11.4.4-1                                                       amd64        CUDA Runtime 11.4 meta-package
ii  cuda-samples-11-4                                   11.4.120-1                                                     amd64        CUDA example applications
ii  cuda-sanitizer-11-4                                 11.4.120-1                                                     amd64        CUDA Sanitizer
ii  cuda-toolkit-11-4                                   11.4.4-1                                                       amd64        CUDA Toolkit 11.4 meta-package
ii  cuda-toolkit-11-4-config-common                     11.4.148-1                                                     all          Common config package for CUDA Toolkit 11.4.
ii  cuda-toolkit-11-config-common                       11.6.55-1                                                      all          Common config package for CUDA Toolkit 11.
ii  cuda-toolkit-config-common                          11.6.55-1                                                      all          Common config package for CUDA Toolkit.
ii  cuda-tools-11-4                                     11.4.4-1                                                       amd64        CUDA Tools meta-package
ii  cuda-visual-tools-11-4                              11.4.4-1                                                       amd64        CUDA visual tools
hi  libcudnn8                                           8.2.2.26-1+cuda11.4                                            amd64        cuDNN runtime libraries
hi  libcudnn8-dev                                       8.2.2.26-1+cuda11.4                                            amd64        cuDNN development libraries and headers
hi  libnvinfer-dev                                      8.2.2-1+cuda11.4                                               amd64        TensorRT development libraries and headers
hi  libnvinfer-plugin-dev                               8.2.2-1+cuda11.4                                               amd64        TensorRT plugin libraries
hi  libnvinfer-plugin8                                  8.2.2-1+cuda11.4                                               amd64        TensorRT plugin libraries
hi  libnvinfer8                                         8.2.2-1+cuda11.4                                               amd64        TensorRT runtime libraries
hi  libnvonnxparsers-dev                                8.2.2-1+cuda11.4                                               amd64        TensorRT ONNX libraries
hi  libnvonnxparsers8                                   8.2.2-1+cuda11.4                                               amd64        TensorRT ONNX libraries
hi  libnvparsers-dev                                    8.2.2-1+cuda11.4                                               amd64        TensorRT parsers libraries
hi  libnvparsers8                                       8.2.2-1+cuda11.4                                               amd64        TensorRT parsers libraries
```

With these versions installed, the setup script failed. :pleading_face: 
```
TASK [autoware.dev_env.tensorrt : Install cuDNN and TensorRT] *****************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"cache_update_time": 1649304301, "cache_updated": true, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"       install 'libcudnn8=8.2.4.15-1+cuda11.4' 'libcudnn8-dev=8.2.4.15-1+cuda11.4' 'libnvinfer8=8.2.4-1+cuda11.4' 'libnvinfer-dev=8.2.4-1+cuda11.4' 'libnvinfer-plugin8=8.2.4-1+cuda11.4' 'libnvinfer-plugin-dev=8.2.4-1+cuda11.4' 'libnvonnxparsers8=8.2.4-1+cuda11.4' 'libnvonnxparsers-dev=8.2.4-1+cuda11.4' 'libnvparsers8=8.2.4-1+cuda11.4' 'libnvparsers-dev=8.2.4-1+cuda11.4' --allow-downgrades' failed: E: Held packages were changed and -y was used without --allow-change-held-packages.\n", "rc": 100, "stderr": "E: Held packages were changed and -y was used without --allow-change-held-packages.\n", "stderr_lines": ["E: Held packages were changed and -y was used without --allow-change-held-packages."], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following held packages will be changed:\n  libcudnn8 libcudnn8-dev libnvinfer-dev libnvinfer-plugin-dev\n  libnvinfer-plugin8 libnvinfer8 libnvonnxparsers-dev libnvonnxparsers8\n  libnvparsers-dev libnvparsers8\nThe following packages will be upgraded:\n  libcudnn8 libcudnn8-dev libnvinfer-dev libnvinfer-plugin-dev\n  libnvinfer-plugin8 libnvinfer8 libnvonnxparsers-dev libnvonnxparsers8\n  libnvparsers-dev libnvparsers8\n10 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "The following held packages will be changed:", "  libcudnn8 libcudnn8-dev libnvinfer-dev libnvinfer-plugin-dev", "  libnvinfer-plugin8 libnvinfer8 libnvonnxparsers-dev libnvonnxparsers8", "  libnvparsers-dev libnvparsers8", "The following packages will be upgraded:", "  libcudnn8 libcudnn8-dev libnvinfer-dev libnvinfer-plugin-dev", "  libnvinfer-plugin8 libnvinfer8 libnvonnxparsers-dev libnvonnxparsers8", "  libnvparsers-dev libnvparsers8", "10 upgraded, 0 newly installed, 0 to remove and 1 not upgraded."]}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=58   changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0
```

**Comment by kenji-miyake** [comment]
`apt-mark unhold` can solve this problem.

```
sudo apt-mark unhold        \
        "cuda*"             \
        "libcudnn*"         \
        "libnvinfer*"       \
        "libnvonnxparsers*" \
        "libnvparsers*"     \
        "nvidia*"
```

```
TASK [autoware.dev_env.tensorrt : Install cuDNN and TensorRT] *****************************************************************************************************************************************************
changed: [localhost]

TASK [autoware.dev_env.tensorrt : Prevent CUDA-related packages from upgrading] ***********************************************************************************************************************************
changed: [localhost] => (item=libcudnn8)
changed: [localhost] => (item=libcudnn8-dev)
changed: [localhost] => (item=libnvinfer8)
changed: [localhost] => (item=libnvinfer-dev)
changed: [localhost] => (item=libnvinfer-plugin8)
changed: [localhost] => (item=libnvinfer-plugin-dev)
changed: [localhost] => (item=libnvonnxparsers8)
changed: [localhost] => (item=libnvonnxparsers-dev)
changed: [localhost] => (item=libnvparsers8)
changed: [localhost] => (item=libnvparsers-dev)

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=60   changed=2    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0

Completed.
```

**Comment by wep21** [comment]
@kenji-miyake Your installed libraries are once held, so I guess you should make allow-change-held-packages option true.
https://github.com/ansible/ansible/pull/73629

**Comment by kenji-miyake** [comment]
@wep21 Oh, it's a good idea. Could you add the option to this PR?

**Comment by wep21** [comment]
@kenji-miyake allow-change-held-packages option will be added in ansible 2.13, but unfortunately the latest version from pip is 2.12.4...
https://github.com/ansible/ansible/blob/66c58449782d94cb5acb01b8d1ef8a207307a496/lib/ansible/modules/apt.py#L116
```
❯ ansible --version
ansible [core 2.12.4]
  config file = None
  configured module search path = ['/home/daisuke/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.8/dist-packages/ansible
  ansible collection location = /home/daisuke/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.8.10 (default, Mar 15 2022, 12:22:08) [GCC 9.4.0]
  jinja version = 3.1.1
  libyaml = True
```

**Comment by kenji-miyake** [comment]
@xmfcx Can we merge this? I've added a troubleshooting manual to update the versions.
https://github.com/autowarefoundation/autoware-documentation/pull/76

And we'll create an issue to upgrade to `ansible 2.13` and add `--allow-change-held-packages`.
-> https://github.com/autowarefoundation/autoware.ai/issues/2430

File: ansible/roles/tensorrt/README.md [hunk 1]
```diff
@@ -17,11 +17,11 @@ For Universe, the `cudnn_version` and `tensorrt_version` variable can also be fo
...
 
-cudnn_version="8.2.2.26-1+cuda11.4"
+cudnn_version="8.2.4.15-1+cuda11.4"
 sudo apt-get install libcudnn8=${cudnn_version} libcudnn8-dev=${cudnn_version}
 sudo apt-mark hold libcudnn8 libcudnn8-dev
 
-tensorrt_version="8.2.2-1+cuda11.4"
+tensorrt_version="8.2.4-1+cuda11.4"
 sudo apt-get install libnvinfer8=${tensorrt_version} libnvonnxparsers8=${tensorrt_version} libnvparsers8=${tensorrt_version} libnvinfer-plugin8=${tensorrt_version} libnvinfer-dev=${tensorrt_version} libnvonnxparsers-dev=${tensorrt_version} libnvparsers-dev=${tensorrt_version} libnvinfer-plugin-dev=${tensorrt_version} python3-libnvinfer=${tensorrt_version}
 sudo apt-mark hold libnvinfer8 libnvonnxparsers8 libnvparsers8 libnvinfer-plugin8 libnvinfer-dev libnvonnxparsers-dev libnvparsers-dev libnvinfer-plugin-dev python3-libnvinfer
```
Comment (xmfcx) [inline ansible/roles/tensorrt/README.md:26]: @wep21 @kenji-miyake I ran these lines and it upgraded without issue.

**Review by xmfcx — APPROVED** [review]
I tested the manual installation part and it works without problems. Thanks!

**Comment by kenji-miyake** [comment]
@wep21 @xmfcx @yukkysaito @mitsudome-r We should have checked the ARM versions. :pleading_face: 

https://github.com/autowarefoundation/autoware/runs/6032607872?check_suite_focus=true#step:5:4845
image

**Comment by kenji-miyake** [comment]
@wep21 Could I revert this for the following reasons?

1. ARM build isn't passing.
https://github.com/autowarefoundation/autoware/actions/workflows/docker-build-and-push-self-hosted.yaml

2. Unfortunately, it seems TensorRT for CUDA 11.6 isn't released yet with the network installation. :disappointed: (So we can't upgrade to `11.6`.)

Or should we specify different versions for `amd64` and `arm64`?

**Pull request merged**